bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/firmware/FirmwareUpdateServiceTest.java

```diff
@@ -318,6 +319,39 @@ public void testCancelFirmwareUpdate() {
             verify(handler3, times(1)).cancel();
         });
     }
+    
+    @Test
+    public void testCancelFirmwareUpdateIntheMiddleOfUpdate() {
+        final long stepsTime = 200;
+        final int numberOfSteps = SEQUENCE.length;
+        final AtomicBoolean isUpdateFinished = new AtomicBoolean(false);
+        
+        doAnswer(invocation -> {
+            ProgressCallback progressCallback = (ProgressCallback) invocation.getArguments()[1];
+            progressCallback.defineSequence(SEQUENCE);
+            
+            // Simulate update steps with delay
+            for(int updateStepsCount=0; updateStepsCount < numberOfSteps; updateStepsCount++) {
+                progressCallback.next();
+                Thread.sleep(stepsTime);
+            }
+            progressCallback.success();
+            isUpdateFinished.set(true);
+            return null;
+        }).when(handler1).updateFirmware(any(Firmware.class), any(ProgressCallback.class));
+        
+        // Execute update and cancel it immediately
+        firmwareUpdateService.updateFirmware(THING1_UID, FW112_EN.getUID(), null);
+        firmwareUpdateService.cancelFirmwareUpdate(THING1_UID);
+        
+        // Be sure that the cancel is executed before the completion of the update
+        waitForAssert(() -> {
+            verify(handler1, times(1)).cancel();
+        },stepsTime * numberOfSteps,50);
+        
+        // Check that the update is not finished even if the cancel has been called within the timeout
+        assertThat("Cancel should stop long running updates", isUpdateFinished.get(), is(false));
```

Comment:
Please drop the explanation string here - it makes the tests harder to read because it states exactly the opposite of what your are asserting here and hence completely neglects the "fluent" API of the hamcrest matchers.

Reply:
Okay, got it - no explanations :) We read code :) :)